Src/VimSpecific/Implementation/Selection/MultiSelectionUtil.cs

```diff
@@ -0,0 +1,83 @@
+﻿#if VS_SPECIFIC_2015 || VS_SPECIFIC_2017
+#else
+
+using Microsoft.VisualStudio.Text.Editor;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Vim;
+using Vim.VisualStudio.Specific;
+
+namespace Vim.Specific.Implementation.Selection
+{
+    [Export(typeof(ISelectionUtil))]
+    [Export(typeof(IVimSpecificService))]
+    class MultiSelectionUtil : VimSpecificService, ISelectionUtil
+    {
+        [ImportingConstructor]
+        internal MultiSelectionUtil(Lazy<IVimHost> vimHost)
+            : base(vimHost)
+        {
+        }
+
+        bool ISelectionUtil.IsMultiSelectionSupported => true;
+
+        IEnumerable<SelectedSpan> ISelectionUtil.GetSelectedSpans(ITextView textView)
+        {
+            var broker = textView.GetMultiSelectionBroker();
+            var primarySelection = broker.PrimarySelection;
+            if (textView.Selection.Mode != TextSelectionMode.Stream)
+            {
+                return new[] { GetSelectedSpan(primarySelection) };
+            }
+            var secondarySelections = broker.AllSelections
+                .Where(span => span != primarySelection)
+                .Select(selection => GetSelectedSpan(selection));
+            return new[] { GetSelectedSpan(primarySelection) }.Concat(secondarySelections);

```

Comment:
Wow that is unintuitive for `AllSelections` to not include the `PrimarySelection`. Generally I read "all" to mean ... all. Seems like that API would be much better named to be `AllSecondarySelections`. 

This code is definitely correct, but that API surprising.

Reply:
The new multi-selection broker is basically undocumented, at least as far as nuance is concerned. Without documentation, it is impossible to know if we can reliably count on the first selection of all selections to be the primary selection. For example, they could could be auto-sorted by buffer position. If that is the case, then the primary selection might be the third selection.

The contract of the `SelectionUtil` API that we expose to VsVim ensures that the the first selected span is always the primary selection. With better documentation we might be able to make this method more efficient and use less memory, but until then, I chose the safest route.

Reply:
My original comment was wrong: `AllSelections` does include `PrimarySelection`. 

I agree the documentation is lacking here. Eventually I just pulled up the type under ILSpy to figure out how it works. 

> Without documentation, it is impossible to know if we can reliably count on the first selection of all selections to be the primary selection. 

The code seems to allow for the exact opposite: the primary selection being the non-first one. This is based on my reading of `SetSelectionRange`. That takes an `IEnumerable<Selection>` to select and `Selection` which is to be made the primary. The resulting order of `AllSelections` is essentially the `IEnumerable<Selection>` sorted by the `InsertionPoint` value. The `PrimarySelection` value though is just whichever value is reference equal to the `primary` parameter.

Reply:
Thanks for sussing that out.